{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding novel insertions\n",
    "\n",
    "Environment: bioconda's `sambamba samtools pigz parallel bxtools blast snakemake spades bioawk` + `supernova` and `ema` (installed in extras)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLES = ['NA12878_WGS', 'COLO829-10x']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run longranger to get `NA12878_WGS_v2_phased_possorted_bam.bam` or `COLO829-10x_phased_possorted_bam.bam`\n",
    "\n",
    "```\n",
    "cd /data/cephfs/punim0010/projects/Saveliev_10X/COLO829-10x/insertions\n",
    "```\n",
    "\n",
    "Extract unmapped reads or reads with unmapped mate:\n",
    "\n",
    "```\n",
    "mkdir unmapped_or_mate_is_unmapped\n",
    "\n",
    "NA12878:\n",
    "sambamba view -f bam -F \"unmapped or mate_is_unmapped\" -t 20 /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/NA12878_WGS_v2_phased_possorted_bam.bam | sambamba sort -n -t {threads} -m 100G /dev/stdin --show-progress > unmapped_or_mate_is_unmapped/NA12878_WGS.bam\n",
    "\n",
    "COLO829:\n",
    "sambamba view -f bam -F \"unmapped or mate_is_unmapped\" -t 20 COLO829-10x_phased_possorted_bam.bam | sambamba sort -n -t {threads} -m 100G /dev/stdin --show-progress > unmapped_or_mate_is_unmapped/COLO829-10x.bam\n",
    "```\n",
    "\n",
    "Split by BX tag:\n",
    "\n",
    "```\n",
    "cd unmapped_or_mate_is_unmapped\n",
    "mkdir bx_split\n",
    "\n",
    "bxtools split NA12878_WGS.bam -a bx_split/NA12878_WGS -m 1000 > bx_split/NA12878_WGS_bx_counts.txt\n",
    "\n",
    "bxtools split COLO829-10x.bam -a bx_split/COLO829-10x -m 1000 > bx_split/COLO829-10x_bx_counts.txt\n",
    "```\n",
    "\n",
    "Extract tags with at least 1000 reads:\n",
    "\n",
    "```\n",
    "sort -k 2 -r -n NA12878_WGS_bx_counts.txt | py -x \"x if int(x.split()[1]) >= 1000 else None\" > NA12878_WGS_bx_counts_cnt1000.txt\n",
    "\n",
    "sort -k 2 -r -n COLO829-10x_bx_counts.txt | py -x \"x if int(x.split()[1]) >= 1000 else None\" > COLO829-10x_bx_counts_cnt1000.txt\n",
    "```\n",
    "\n",
    "Count all reads in full BAM by BX:\n",
    "\n",
    "```\n",
    "bxtools stats /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/NA12878_WGS_v2_phased_possorted_bam.bam > /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/qc/NA12878_WGS_v2_phased_possorted_bam.bxstats.txt\n",
    "\n",
    "bxtools stats /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/NA12878_WGS_v2_phased_possorted_bam.bam > /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/qc/NA12878_WGS_v2_phased_possorted_bam.bxstats.txt\n",
    "\n",
    "```\n",
    "\n",
    "Extract BX with many unmapped:\n",
    "\n",
    "```\n",
    "grep -f <(cut -f1 NA12878_WGS_bx_counts_cnt1000.txt) /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/qc/NA12878_WGS_v2_phased_possorted_bam.bxstats.txt > NA12878_WGS_full_bx_stats_cnt1000.txt\n",
    "```\n",
    "\n",
    "------------------------\n",
    "Optional:\n",
    "\n",
    "Split full BAM by BX tag:\n",
    "\n",
    "```\n",
    "bxtools split NA12878_WGS_v2_phased_possorted_bam.bam -a bx_split/NA12878_WGS -m 1000 > bx_split/NA12878_WGS_bx_counts.txt\n",
    "```\n",
    "\n",
    "Convert unmapped reads to fastq\n",
    "\n",
    "```\n",
    "bedtools bamtofastq -i - -fq {output.fq1} -fq2 {output.fq2}'\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as pt\n",
    "from pprint import pprint\n",
    "from collections import namedtuple\n",
    "\n",
    "bxs = dict()\n",
    "\n",
    "with open('NA12878_WGS_bx_counts_unmapped_or_mate_top.txt') as f:\n",
    "    for l in f:\n",
    "        l = l.strip()\n",
    "        if l:\n",
    "            bx, cnt = l.split()\n",
    "            cnt = int(cnt)\n",
    "            if cnt >= 1000:\n",
    "                bxs[bx] = {'Or mate': int(cnt)}\n",
    "\n",
    "with open('NA12878_WGS_bx_counts_unmapped_top.txt') as f:\n",
    "    for l in f:\n",
    "        l = l.strip()\n",
    "        if l:\n",
    "            bx, cnt = l.split()\n",
    "            if bx in bxs:\n",
    "                bxs[bx]['Unaln'] = int(cnt)\n",
    "    \n",
    "# counts = [c for bx, c in bx_cnt.items() if c > 500]\n",
    "# print(len(counts))\n",
    "# pt.plot(counts)\n",
    "\n",
    "with open('NA12878_WGS_full_bx_stats_cnt1000.txt') as f:\n",
    "    for l in f:\n",
    "        l = l.strip()\n",
    "        if l:\n",
    "            bx, cnt, mis, mmq, mas = l.split()\n",
    "            if bx in bxs:\n",
    "                bxs[bx]['Aln'] = int(cnt)\n",
    "                bxs[bx]['IS'] = int(mis)\n",
    "                bxs[bx]['MQ'] = int(mmq)\n",
    "                bxs[bx]['AS'] = int(mas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\n",
      "                     Aln  Unaln | Mate unaln  Unaln  Aln & Mate unaln   IS  MQ\n",
      "CATGTAGAGTCGTACT-1  4037              94.70% 94.43%                11  237  60\n",
      "GCCTGTTAGCCCGTGT-1  1342              86.44% 85.77%                 9  348  60\n",
      "ACTGTAGGTTCGTAAC-1  1647              84.03% 83.79%                 4  347  60\n",
      "CATGTAGGTAGATGTA-1  1250              84.00% 83.28%                 9   19  19\n",
      "GCCTGTTGTTCGCGGT-1  1921              73.30% 72.98%                 6  348  60\n",
      "CATGTAGAGAGTCTCT-1  1828              72.87% 72.10%                14   20  15\n",
      "GCCTGTTGTCTTACCC-1  1827              72.03% 71.65%                 7  374  60\n",
      "GCCTGTTAGTCGAGTG-1  1515              69.37% 68.84%                 8  349  60\n",
      "TCAGTAGGTCATGTAC-1  2070              69.47% 68.41%                22  304  60\n",
      "GCCTGTTAGCTCCGTG-1  1714              68.14% 67.74%                 7  368  60\n",
      "ACTGTAGAGGTAAACT-1  1507              67.09% 66.42%                10   20  19\n",
      "GCCTGTTAGGCCCTCA-1  3324              63.96% 63.45%                17  308  60\n",
      "CATGTAGGTTGATCAC-1  1649              63.61% 62.77%                14  315  60\n",
      "GCCTGTTGTTTCGTTT-1  1933              63.68% 62.44%                24  322  60\n",
      "GCCTGTTAGTCGGCTC-1  2126              61.90% 61.43%                10  339  60\n",
      "GCCTGTTGTTTGGCTA-1  1921              64.29% 56.95%               141  314  60\n",
      "TGCGCAGGTACTGTAC-1  2327              56.81% 55.82%                23  334  60\n",
      "GCCTGTTGTCCGGGTT-1  2053              52.70% 50.02%                55  288  49\n",
      "GCCTGTTGTACGTTTG-1  2316              50.69% 50.00%                16  286  60\n",
      "AAGCCCGCATCCCACT-1  2524              41.32% 37.08%               107  264  20\n",
      "GTGCCTTCAGATCTGT-1  7461              21.61% 20.68%                69  336  60\n",
      "GTTGCAACAGATCTGT-1  5943              19.60% 15.01%               273  329  60\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.DataFrame.from_dict(bxs, orient='index')\n",
    "print(len(df))\n",
    "df['Aln & Mate unaln'] = df['Unaln | Mate unaln'] - df['Unaln']\n",
    "\n",
    "df['Unaln | Mate unaln'] = 100.0 * df['Unaln | Mate unaln'] / df['Aln']\n",
    "df['Unaln'] = 100.0 * df['Unaln'] / df['Aln']\n",
    "#df['Aln & Mate unaln'] = 100.0 * df['Aln & Mate unaln'] / df['Aln']\n",
    "\n",
    "df = df.sort_values(by='Unaln', ascending=False)\n",
    "\n",
    "pd.options.display.float_format = '{:.2f}%'.format\n",
    "#print('\\n'.join(df.index))\n",
    "print(df[['Aln', 'Unaln | Mate unaln', 'Unaln', 'Aln & Mate unaln', 'IS', 'MQ']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each tag, extract mapped reads with unmapped mates:\n",
    "\n",
    "```python\n",
    "!cd /data/cephfs/punim0010/projects/Saveliev_10X/COLO829-10x/insertions/unmapped_or_mate_is_unmapped/split_by_bx/\n",
    "\n",
    "for tag in df.index:\n",
    "    !sambamba view -F \"not unmapped and mapping_quality >= 60\" NA12878_WGS.{tag}-1.bam -f bam | samtools sort > NA12878_WGS.{tag}-1.mapped.bam\n",
    "```\n",
    "\n",
    "Manually reviewing the reads, we can find one BX with mapped reads spanning consistently more or less one region (possibly breakpoint?)\n",
    "\n",
    "```\n",
    "samtools view NA12878_WGS.GCCTGTTGTTTCGTTT-1.mapped.bam\n",
    "\n",
    "ST-E00273:259:H7WY3ALXX:1:1217:21207:15479  121 chr2  71844818  60  73M1D11M2I30M4D12M  chr7  128046340  0\n",
    "ST-E00273:259:H7WY3ALXX:1:1220:8156:43677    73 chr2  71844971  60  47M6I7M2I66M        chr10    654413  0\n",
    "ST-E00273:259:H7WY3ALXX:2:1118:11373:44152   73 chr2  71845284  60  52M13I2M1D61M       chr21  44035538  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2121:15006:67814   89 chr2  71845342  60  14S54M7I53M         =      33141434  0\n",
    "ST-E00273:259:H7WY3ALXX:3:2117:1499:37664    89 chr2  71845402  60  77M2I22M1I17M1D9M   =      71845700  0\n",
    "ST-E00273:259:H7WY3ALXX:1:1112:18913:57038  169 chr2  71845409  60  7M2I66M76S          *             0  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2124:12134:1977    73 chr2  71845440  60  79M2D21M1D24M4S     chr20  25048324  0\n",
    "ST-E00273:259:H7WY3ALXX:2:1106:10602:14828  153 chr2  71845521  60  30S25M2I40M1D13M41S *             0  0\n",
    "ST-E00273:259:H7WY3ALXX:2:1105:1458:67937    89 chr2  71845525  60  35S93M              chr4   72052386  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2203:29985:65001   89 chr2  71845588  60  9S78M2D29M2D12M     =      71844822  0\n",
    "ST-E00273:259:H7WY3ALXX:3:2207:11921:48230  153 chr2  71845590  60  75S54M7I15M         *             0  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2120:18832:35801   89 chr2  71845617  60  46M2D79M3S          =      71845704  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2204:30888:68430   73 chr2  71845848  60  6S86M36S            hs37d5 31586471  0\n",
    "ST-E00273:259:H7WY3ALXX:1:2204:18913:24198   73 chr4  53065963  60  23M4I19M3D38M44S    chr15  79384471  0\n",
    "ST-E00273:259:H7WY3ALXX:1:1104:28270:7181   121 chr14 32925056  60  128M                chr1   30986160  0\n",
    "ST-E00273:259:H7WY3ALXX:2:1101:22110:57899   73 chr14 32966430  60  128M                chr13  52303476  0\n",
    "ST-E00273:259:H7WY3ALXX:2:2103:30452:36311   89 chr14 33028682  60  128M                chr18  71871932  0\n",
    "ST-E00273:259:H7WY3ALXX:3:2201:23977:47316  105 chr21 28080947  60  128M                chr5   77330570  0\n",
    "```\n",
    "\n",
    "Now get all reads for that tag to see if they localized at the same region, then we'll see it in IGV:\n",
    "\n",
    "```\n",
    "cd /data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x\n",
    "bamtools filter -tag BX:GCCTGTTGTTTCGTTT-1 -in NA12878_WGS_v2_phased_possorted_bam.bam -out NA12878_WGS_v2_phased_possorted_bam-GCCTGTTGTTTCGTTT-1.bam\n",
    "```\n",
    "\n",
    "Extract that region to download locally on IGV:\n",
    "\n",
    "```\n",
    "sambamba slice NA12878_WGS_v2_phased_possorted_bam.bam chr2:71844818-71847848 -o NA12878_WGS_v2_phased_possorted_chr2_breakpoint.bam\n",
    "```\n",
    "\n",
    "Extracting this one to for comparison:\n",
    "\n",
    "```\n",
    "sambamba slice NA12878_WGS_v2_phased_possorted_bam.bam chr19:33788840-33791956 -o NA12878_WGS_v2_phased_possorted_CEBPA.bam\n",
    "\n",
    "```\n",
    "\n",
    "In the large SVs file downloaded from the 10x website, we can find those SVs called with LOWQ: `/data/cephfs/punim0010/projects/Saveliev_10X/NA12878-10x/longranger/downloaded/NA12878_WGS_v2_large_svs.vcf.gz`\n",
    "\n",
    "```\n",
    "chr2   72650821  call_2031    A  <INV>                 5  LOWQ  END=72704553;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=53732;PS=71845775;HAP_ALLELIC_FRAC=0.25;ALLELIC_FRAC=0.0200501253133;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=.;MolDel=.\n",
    "chr2   72790392  call_3830    A  <DUP:TANDEM>          3  LOWQ  END=72858227;CIEND=0,1;CIPOS=0,1;SVTYPE=DUP;SVLEN=67835;PS=71845775;HAP_ALLELIC_FRAC=0.230769230769;ALLELIC_FRAC=0.00617283950617;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR\n",
    "chr2   73124465  call_2875    G  <INV>                 4  LOWQ  END=73195439;CIEND=-24,25;CIPOS=-24,25;SVTYPE=INV;SVLEN=70974;PS=71845775;HAP_ALLELIC_FRAC=0.153846153846;ALLELIC_FRAC=0.0128755364807;PAIRS=2;SPLIT=0;WildCov=.;MolTtl=.;MolT\n",
    "chr2   73877193  call_2032    A  <UNK>                 5  LOWQ  END=73940170;CIEND=-8,8;CIPOS=-6,6;SVTYPE=UNK;SVLEN=62977;PS=71845775;HAP_ALLELIC_FRAC=0.172413793103;ALLELIC_FRAC=0.0178173719376;PAIRS=3;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNo\n",
    "chr2   76051830  call_840     C  <INV>                 8  LOWQ  END=76110415;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=58585;PS=71845775;HAP_ALLELIC_FRAC=0.228571428571;ALLELIC_FRAC=0.0220264317181;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=\n",
    "chr2   76213102  call_2033    A  <INV>                 5  LOWQ  END=76259489;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=46387;PS=71845775;HAP_ALLELIC_FRAC=0.192307692308;ALLELIC_FRAC=0.0216216216216;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=\n",
    "chr2   78278376  call_3831    T  <INV>                 3  LOWQ  END=78322534;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=44158;PS=71845775;HAP_ALLELIC_FRAC=0.103448275862;ALLELIC_FRAC=0.0113314447592;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=\n",
    "chr2   78415466  call_2876    A  <UNK>                 4  LOWQ  END=78476352;CIEND=-9,9;CIPOS=-9,9;SVTYPE=UNK;SVLEN=60886;PS=71845775;HAP_ALLELIC_FRAC=0.137931034483;ALLELIC_FRAC=0.0176767676768;PAIRS=3;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNo\n",
    "chr2   78887169  call_3832    G  <INV>                 3  LOWQ  END=78950399;CIEND=0,1;CIPOS=0,1;SVTYPE=INV;SVLEN=63230;PS=71845775;HAP_ALLELIC_FRAC=0.111111111111;ALLELIC_FRAC=0.00973236009732;PAIRS=1;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR\n",
    "chr2   79106744  call_3833    A  <UNK>                 3  LOWQ  END=79159256;CIEND=-30,31;CIPOS=-30,31;SVTYPE=UNK;SVLEN=52512;PS=71845775;HAP_ALLELIC_FRAC=0.2;ALLELIC_FRAC=0.0171568627451;PAIRS=2;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=.;Mol\n",
    "chr2   80313256  call_1487    T  <INV>                 6  LOWQ  END=80353748;CIEND=-15,15;CIPOS=-15,15;SVTYPE=INV;SVLEN=40492;PS=71845775;HAP_ALLELIC_FRAC=0.25;ALLELIC_FRAC=0.0263157894737;PAIRS=2;SPLIT=0;WildCov=.;MolTtl=.;MolTtlNoR=.;Mo\n",
    "```\n",
    "\n",
    "Viewing in IGV: `chr2:71,844,732-71,845,762`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### De novo assembly of unmapped reads with Supernova\n",
    "\n",
    "Extract reads:\n",
    "\n",
    "```\n",
    "bedtools bamtofastq -i COLO829-10x.bam -fq COLO829-10x_S1_L001_R1_001.fastq -fq2 COLO829-10x_S1_L001_R2_001.fastq 2>COLO829-10x.unpaired\n",
    "bedtools bamtofastq -i NA12878_WGS.bam -fq NA12878_WGS_S1_L001_R1_001.fastq -fq2 NA12878_WGS_S1_L001_R2_001.fastq 2>NA12878_WGS.unpaired\n",
    "```\n",
    "\n",
    "Preproc reads:\n",
    "\n",
    "```\n",
    "\n",
    "```\n",
    "\n",
    "Assemble with supernova:\n",
    "\n",
    "```\n",
    "cd /g/data3/gx8/projects/Saveliev_10X/NA12878-10x/insertions/unmapped_or_mate_is_unmapped/assemble\n",
    "qsub run.sh\n",
    "```\n",
    "\n",
    "Fails with error:\n",
    "\n",
    "```\n",
    "[error] The fraction of input reads having valid barcodes is 1.68 pct, whereas the ideal is at least 80 pct.  This condition could have multiple causes including wrong library type, failed library construction and low sequence quality on the barcode bases.  This could have a severe effect on assembly performance, and Supernova has not been tested on data with these properties, so execution will be terminated.\n",
    "\n",
    "We observe only 36.67 pct of bases on read two with quality scores at least 30. Ideally, we expect at least 75 pct. Data quality issues of this type are difficult to diagnose, but might be caused by defects in sequencing reagents or sequencing instrument condition. Use of low quality reads usually reduces assembly quality.\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
